### DISCUSSION TITLE
Facial Recognition re-processes lots of assets

### DISCUSSION THREAD

**Discussion by tbleiker** [immich-app/immich#6571] [Feature Request]
### The bug

When I click "Missing" on the "Facial Recognition" job, thousands of assets are re-processed. When the job finishes and I click "Missing" again, the same amount of assets are processed again. 

### The OS that Immich Server is running on

Docker on Debian Bookworm

### Version of Immich Server

v1.93.3

### Version of Immich Mobile App

v1.93.2

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
-
```

### Your .env content

```Shell
-
```

### Reproduction steps

```bash
1. Open "Administration" and then "Jobs" 
2. Click "Missing" under Facial Recognition
```

**Comment by mertalev** [comment]
This isn't a bug. "Missing" queues faces with no associated person. It needs to process all of these to see if any are eligible to be grouped into people.

**Comment by tbleiker** [comment]
I see. This is due to "min recognized faces" set to >1?

This is a little unsatisfactory, because every time I take a picture, the server now processes thousands of images (which takes roughly 15min now, avg. 50% cpu for all my 4 cores). And the number of assets to process potentially grows with every image I take...

**Comment by mertalev** [comment]
It's possible to reduce the number of faces processed by only checking the faces of new assets, but this would lower the quality of the clustering algorithm.

If you set min faces to 1, the behavior will resemble the old algorithm: all faces will be associated with a person and the clustering for new assets will be quick. The other side of that is that the quality will also be closer to the old algorithm (but still somewhat better).

**Comment by tbleiker** [comment]
Thx for the fast response.

Maybe, "only checking the faces of new assets" can be provided as option and the user can decide. 

For my assets, I cannot see an overall improvement of the new clustering algorithm. Therefore, I would prefer speed (and less ressources) over quality 😄 

Edit: Nevertheless, I would like to still use "min faces >1"...

**Comment by tbleiker** [comment]
Oh, and I realised that the facial recognition job is triggered as well for a new image without detected faces.

Edit: rephrasing...

**Comment by mertalev** [comment]
That's a good point. We can optimize that to only trigger facial recognition when a face is detected.

**Comment by tbleiker** [comment]
What do you think, are there any other optimization options? Should we open a discussion topic for this? Or discuss it in Discord?

E.g.: I would expect, that the facial recognition for a new picture with e.g. two faces (which are already clusterd) is fast - given that their points are right in the middle of their already known cluster.

**Reply by mertalev** [reply]
A solution could be to only check new assets if facial recognition is run due to an uploaded image. The quality cost of this is in not creating new people or not assigning a person to a face in some cases. But the clustering that does happen isn't affected. Manual "missing" jobs would maintain the current behavior, and we could have a nightly "missing" job to do the more thorough check.

**Reply by tbleiker** [reply]
I am not sure if I correctly understand "only check new assets if facial recognition is run due to an uploaded image". --> A new picture gets taken, uploaded and face detection is run. What's next?

Is this possible: Representing point of a detected face ...
- is in the midst of a known cluster (or surround by core points of the same cluster within the given distance): "fast" tagging
- otherwise: postbone more thorough check 

Would this allow to "quickly" tag known persons in a picture and add the other persons after the thorough check?

**Reply by mertalev** [reply]
I mean that instead of queueing all faces with no person, it would only queue the detected faces of the new assets.

This is effectively the "fast path" option:

* If the face is a "core" face (i.e. there are min faces - 1 faces around it) and a person exists for it, clustering works as normal.
* If it's a core face and there's no person for it, a person is created but the other faces aren't clustered unless they're in the same batch of new faces.
* If it's a non-core face and there's a similar face with a person, clustering works as normal.
* If it's a non-core face and there's no person for it, then it won't be clustered.

**Reply by tbleiker** [reply]
I haven't yet thought that completly through (with my limited knowledge... 😄) - but that sounds promising...!

**Reply by mertalev** [reply]
This change will be in the next release through PR #6634

**Reply by tbleiker** [reply]
Thx a lot! Do you need some additional testing before the next release? Maybe I can find some time in the next few days...

**Reply by mertalev** [reply]
That'd be nice! You can try it by changing the image tag in `immich-microservices` to `pr-6634` (probably good to set the server to this too, but not required)

**Reply by tbleiker** [reply]
I wanted to give it a try... but there is no tag `pr-6634` for `immich-microservices` (`immich-server`), but for `immich-machine-learning` and only with `cuda`, `openvino` or `armnn`.

I am tempted to use `main` for all images... but I am afraid to break something :stuck_out_tongue:
